In [1]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

In [3]:
print(df.shape)


(244, 7)


In [4]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='str')

In [5]:
df.dtypes

total_bill     float64
tip            float64
sex           category
smoker        category
day           category
time          category
size             int64
dtype: object

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB


In [7]:
df.describe()

,total_bill,tip,size
count,244.000000,244.000000,244.000000
mean,19.785943,2.998279,2.569672
std,8.902412,1.383638,0.951100
min,3.070000,1.000000,1.000000
25%,13.347500,2.000000,2.000000
50%,17.795000,2.900000,2.000000
75%,24.127500,3.562500,3.000000
max,50.810000,10.000000,6.000000


In [8]:
df["total_bill"].mean()

np.float64(19.78594262295082)

In [9]:
df["tip"].max()

np.float64(10.0)

In [12]:
df["tip"]
#ovo vraca series podataka 1D

0      1.01
1      1.66
2      3.50
3      3.31
4      3.61
       ... 
239    5.92
240    2.00
241    2.00
242    1.75
243    3.00
Name: tip, Length: 244, dtype: float64

In [13]:
df[["tip"]]
#ovo vraca dataframe poidataka 2D

,tip
0,1.01
1,1.66
2,3.50
3,3.31
4,3.61
...,...
239,5.92
240,2.00
241,2.00
242,1.75


In [14]:
df[["tip", "total_bill"]]

,tip,total_bill
0,1.01,16.99
1,1.66,10.34
2,3.50,21.01
3,3.31,23.68
4,3.61,24.59
...,...,...
239,5.92,29.03
240,2.00,27.18
241,2.00,22.67
242,1.75,17.82


### Loc gleda indekse kao tekst pa ukljucuje i gornju vrednost dok iloc gleda kao indekse sto gleda python kod slicinga pa gornja granica nije ukljucena

In [15]:
df.iloc[0:3]

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3


In [16]:
df.loc[0:3]

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2


### Loc radi sa oznakama a iloc sa pozicijama tako da kada se filtriraju podaci ta 2 vise nece raditi isto

In [19]:
print(df["tip"].loc[0])

1.01


In [20]:
print(df["tip"].iloc[0])

1.01


In [21]:
df[df["total_bill"] > 40]

,total_bill,tip,sex,smoker,day,time,size
59,48.27,6.73,Male,No,Sat,Dinner,4
95,40.17,4.73,Male,Yes,Fri,Dinner,4
102,44.30,2.50,Female,Yes,Sat,Dinner,3
142,41.19,5.00,Male,No,Thur,Lunch,5
156,48.17,5.00,Male,No,Sun,Dinner,6
170,50.81,10.00,Male,Yes,Sat,Dinner,3
182,45.35,3.50,Male,Yes,Sun,Dinner,3
184,40.55,3.00,Male,Yes,Sun,Dinner,2
197,43.11,5.00,Female,Yes,Thur,Lunch,4
212,48.33,9.00,Male,No,Sat,Dinner,4


In [27]:
df[(df["total_bill"] > 30) & (df["day"] == "Sun") ]

#KADA IMA VISE USLOVA MORAJU DA IDU OVAKO U OBICNE ZAGRADE SVAKI POSEBNO

,total_bill,tip,sex,smoker,day,time,size
11,35.26,5.00,Female,No,Sun,Dinner,4
44,30.40,5.60,Male,No,Sun,Dinner,4
47,32.40,6.00,Male,No,Sun,Dinner,4
52,34.81,5.20,Female,No,Sun,Dinner,4
112,38.07,4.00,Male,No,Sun,Dinner,3
156,48.17,5.00,Male,No,Sun,Dinner,6
167,31.71,4.50,Male,No,Sun,Dinner,4
173,31.85,3.18,Male,Yes,Sun,Dinner,2
175,32.90,3.11,Male,Yes,Sun,Dinner,2
179,34.63,3.55,Male,Yes,Sun,Dinner,2


### Za ove agregacije i to najlakse je da se zamisli sta bi islo u zagradi pri izdvajanju uglastoj pa onda da too bude maska a nad maskom primenjujemo mean sum max itd, inace ke jaska u sustini true false tako da moze procenat odatle da se vadi kad se trazi nesto a sum prebrojava ustvari ono sto smo zadali koliko ima takvih kolona

In [40]:
maska = df["smoker"] == "Yes"
maska.mean()

np.float64(0.38114754098360654)

In [42]:
maska = df["time"] == "Lunch"
maska.sum()

np.int64(68)

In [45]:
print((df["time"] == "Lunch").sum())

68


In [46]:
df.time

0      Dinner
1      Dinner
2      Dinner
3      Dinner
4      Dinner
        ...  
239    Dinner
240    Dinner
241    Dinner
242    Dinner
243    Dinner
Name: time, Length: 244, dtype: category
Categories (2, str): ['Lunch', 'Dinner']

In [47]:
df["time"]

0      Dinner
1      Dinner
2      Dinner
3      Dinner
4      Dinner
        ...  
239    Dinner
240    Dinner
241    Dinner
242    Dinner
243    Dinner
Name: time, Length: 244, dtype: category
Categories (2, str): ['Lunch', 'Dinner']

### isin ima funkciju kao sto je ovaj uslov znaci da vrednosti spadaju u skup vrednosti koji je u listi, nije od 1 do 6 nego tacno 1 ili tacno 6

In [56]:
((df["size"] == 1) | (df["size"] == 6)).sum()

np.int64(8)

In [52]:
df["size"].isin([1,6]).sum()

np.int64(8)

In [60]:
df["tip_pct"] = df["tip"]/df["total_bill"]*100

In [63]:
print(df)

     total_bill   tip     sex smoker   day    time  size    tip_pct
0         16.99  1.01  Female     No   Sun  Dinner     2   5.944673
1         10.34  1.66    Male     No   Sun  Dinner     3  16.054159
2         21.01  3.50    Male     No   Sun  Dinner     3  16.658734
3         23.68  3.31    Male     No   Sun  Dinner     2  13.978041
4         24.59  3.61  Female     No   Sun  Dinner     4  14.680765
..          ...   ...     ...    ...   ...     ...   ...        ...
239       29.03  5.92    Male     No   Sat  Dinner     3  20.392697
240       27.18  2.00  Female    Yes   Sat  Dinner     2   7.358352
241       22.67  2.00    Male    Yes   Sat  Dinner     2   8.822232
242       17.82  1.75    Male     No   Sat  Dinner     2   9.820426
243       18.78  3.00  Female     No  Thur  Dinner     2  15.974441

[244 rows x 8 columns]


In [64]:
df.tip_pct.mean()

np.float64(16.08025817225047)

In [68]:
df["po_osobi"] = df.total_bill/df["size"]
print(df)

     total_bill   tip     sex smoker   day    time  size    tip_pct   po_osobi
0         16.99  1.01  Female     No   Sun  Dinner     2   5.944673   8.495000
1         10.34  1.66    Male     No   Sun  Dinner     3  16.054159   3.446667
2         21.01  3.50    Male     No   Sun  Dinner     3  16.658734   7.003333
3         23.68  3.31    Male     No   Sun  Dinner     2  13.978041  11.840000
4         24.59  3.61  Female     No   Sun  Dinner     4  14.680765   6.147500
..          ...   ...     ...    ...   ...     ...   ...        ...        ...
239       29.03  5.92    Male     No   Sat  Dinner     3  20.392697   9.676667
240       27.18  2.00  Female    Yes   Sat  Dinner     2   7.358352  13.590000
241       22.67  2.00    Male    Yes   Sat  Dinner     2   8.822232  11.335000
242       17.82  1.75    Male     No   Sat  Dinner     2   9.820426   8.910000
243       18.78  3.00  Female     No  Thur  Dinner     2  15.974441   9.390000

[244 rows x 9 columns]


In [70]:
df["velika_grupa"] = df["size"] >= 4
print(df)

     total_bill   tip     sex smoker   day    time  size    tip_pct  \
0         16.99  1.01  Female     No   Sun  Dinner     2   5.944673   
1         10.34  1.66    Male     No   Sun  Dinner     3  16.054159   
2         21.01  3.50    Male     No   Sun  Dinner     3  16.658734   
3         23.68  3.31    Male     No   Sun  Dinner     2  13.978041   
4         24.59  3.61  Female     No   Sun  Dinner     4  14.680765   
..          ...   ...     ...    ...   ...     ...   ...        ...   
239       29.03  5.92    Male     No   Sat  Dinner     3  20.392697   
240       27.18  2.00  Female    Yes   Sat  Dinner     2   7.358352   
241       22.67  2.00    Male    Yes   Sat  Dinner     2   8.822232   
242       17.82  1.75    Male     No   Sat  Dinner     2   9.820426   
243       18.78  3.00  Female     No  Thur  Dinner     2  15.974441   

      po_osobi  velika_grupa  
0     8.495000         False  
1     3.446667         False  
2     7.003333         False  
3    11.840000         

In [71]:
df["velika_grupa"].sum()


np.int64(46)

### Ispod ce da bude malo testiranja sortiranja posto to prvi put vidim

In [74]:
df.sort_values('tip')
#PO DEFAULTU JE ASCENDING

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
92,5.75,1.00,Female,Yes,Fri,Dinner,2,17.391304,2.875000,False
111,7.25,1.00,Female,No,Sat,Dinner,1,13.793103,7.250000,False
67,3.07,1.00,Female,Yes,Sat,Dinner,1,32.573290,3.070000,False
236,12.60,1.00,Male,Yes,Sat,Dinner,2,7.936508,6.300000,False
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673,8.495000,False
...,...,...,...,...,...,...,...,...,...,...
141,34.30,6.70,Male,No,Thur,Lunch,6,19.533528,5.716667,True
59,48.27,6.73,Male,No,Sat,Dinner,4,13.942407,12.067500,True
23,39.42,7.58,Male,No,Sat,Dinner,4,19.228818,9.855000,True
212,48.33,9.00,Male,No,Sat,Dinner,4,18.621974,12.082500,True


In [76]:
df.sort_values('tip', ascending = False)

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
170,50.81,10.00,Male,Yes,Sat,Dinner,3,19.681165,16.936667,False
212,48.33,9.00,Male,No,Sat,Dinner,4,18.621974,12.082500,True
23,39.42,7.58,Male,No,Sat,Dinner,4,19.228818,9.855000,True
59,48.27,6.73,Male,No,Sat,Dinner,4,13.942407,12.067500,True
141,34.30,6.70,Male,No,Thur,Lunch,6,19.533528,5.716667,True
...,...,...,...,...,...,...,...,...,...,...
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673,8.495000,False
111,7.25,1.00,Female,No,Sat,Dinner,1,13.793103,7.250000,False
92,5.75,1.00,Female,Yes,Fri,Dinner,2,17.391304,2.875000,False
67,3.07,1.00,Female,Yes,Sat,Dinner,1,32.573290,3.070000,False


**Ovde ispod je sad prikazano kako se sortira po vise kolona onim redom kojim su navedene i za svaki biramo jel hocemo opadajuce ili raastuce a zatim sa head i tail biramo kraj ili pocetak mada logicno je head ako smo izabrali sortiranje i na kraju u [[]] pisemo koje kolone hocemo da izvucemo da ne bi izvlacili sve**

In [82]:
df.sort_values(['total_bill', 'tip'], ascending = [True, False]).head(3)[["sex"]]

,sex
67,Female
92,Female
172,Male


**Nastavljamo sad sa zadacima koje je claude smislio**

In [84]:
df.sort_values('total_bill', ascending=False)[['tip', 'total_bill']] #koji se redosled stavi u ove 2 uglaste tako i vrati rezultat

,tip,total_bill
170,10.00,50.81
212,9.00,48.33
59,6.73,48.27
156,5.00,48.17
182,3.50,45.35
...,...,...
149,2.00,7.51
111,1.00,7.25
172,5.15,7.25
92,1.00,5.75


In [85]:
df['tip_pct'].idxmax()

172

In [87]:
df.loc[[df['tip_pct'].idxmax()]]

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
172,7.25,5.15,Male,Yes,Sun,Dinner,2,71.034483,3.625,False


In [94]:
df = pd.concat([df, df.loc[[df['tip_pct'].idxmax()]]])

In [95]:
df.tail(5)

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
240,27.18,2.00,Female,Yes,Sat,Dinner,2,7.358352,13.590,False
241,22.67,2.00,Male,Yes,Sat,Dinner,2,8.822232,11.335,False
242,17.82,1.75,Male,No,Sat,Dinner,2,9.820426,8.910,False
243,18.78,3.00,Female,No,Thur,Dinner,2,15.974441,9.390,False
172,7.25,5.15,Male,Yes,Sun,Dinner,2,71.034483,3.625,False


In [99]:
df[170:180]

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
170,50.81,10.00,Male,Yes,Sat,Dinner,3,19.681165,16.936667,False
171,15.81,3.16,Male,Yes,Sat,Dinner,2,19.987350,7.905000,False
172,7.25,5.15,Male,Yes,Sun,Dinner,2,71.034483,3.625000,False
173,31.85,3.18,Male,Yes,Sun,Dinner,2,9.984301,15.925000,False
174,16.82,4.00,Male,Yes,Sun,Dinner,2,23.781213,8.410000,False
175,32.90,3.11,Male,Yes,Sun,Dinner,2,9.452888,16.450000,False
176,17.89,2.00,Male,Yes,Sun,Dinner,2,11.179430,8.945000,False
177,14.48,2.00,Male,Yes,Sun,Dinner,2,13.812155,7.240000,False
178,9.60,4.00,Female,Yes,Sun,Dinner,2,41.666667,4.800000,False
179,34.63,3.55,Male,Yes,Sun,Dinner,2,10.251227,17.315000,False


In [100]:
df.sort_values('tip_pct')

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
237,32.83,1.17,Male,Yes,Sat,Dinner,2,3.563814,16.415000,False
102,44.30,2.50,Female,Yes,Sat,Dinner,3,5.643341,14.766667,False
57,26.41,1.50,Female,No,Sat,Dinner,2,5.679667,13.205000,False
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673,8.495000,False
187,30.46,2.00,Male,Yes,Sun,Dinner,5,6.565988,6.092000,True
...,...,...,...,...,...,...,...,...,...,...
232,11.61,3.39,Male,No,Sat,Dinner,2,29.198966,5.805000,False
67,3.07,1.00,Female,Yes,Sat,Dinner,1,32.573290,3.070000,False
178,9.60,4.00,Female,Yes,Sun,Dinner,2,41.666667,4.800000,False
172,7.25,5.15,Male,Yes,Sun,Dinner,2,71.034483,3.625000,False


In [101]:
df = df.drop_duplicates()

In [102]:
print(df)

     total_bill   tip     sex smoker   day    time  size    tip_pct  \
0         16.99  1.01  Female     No   Sun  Dinner     2   5.944673   
1         10.34  1.66    Male     No   Sun  Dinner     3  16.054159   
2         21.01  3.50    Male     No   Sun  Dinner     3  16.658734   
3         23.68  3.31    Male     No   Sun  Dinner     2  13.978041   
4         24.59  3.61  Female     No   Sun  Dinner     4  14.680765   
..          ...   ...     ...    ...   ...     ...   ...        ...   
239       29.03  5.92    Male     No   Sat  Dinner     3  20.392697   
240       27.18  2.00  Female    Yes   Sat  Dinner     2   7.358352   
241       22.67  2.00    Male    Yes   Sat  Dinner     2   8.822232   
242       17.82  1.75    Male     No   Sat  Dinner     2   9.820426   
243       18.78  3.00  Female     No  Thur  Dinner     2  15.974441   

      po_osobi  velika_grupa  
0     8.495000         False  
1     3.446667         False  
2     7.003333         False  
3    11.840000         

In [103]:
df.sort_values(["day","total_bill"], ascending=[True, False])

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
197,43.11,5.00,Female,Yes,Thur,Lunch,4,11.598237,10.777500,True
142,41.19,5.00,Male,No,Thur,Lunch,5,12.138869,8.238000,True
85,34.83,5.17,Female,No,Thur,Lunch,4,14.843526,8.707500,True
141,34.30,6.70,Male,No,Thur,Lunch,6,19.533528,5.716667,True
83,32.68,5.00,Male,Yes,Thur,Lunch,2,15.299878,16.340000,False
...,...,...,...,...,...,...,...,...,...,...
53,9.94,1.56,Male,No,Sun,Dinner,2,15.694165,4.970000,False
43,9.68,1.32,Male,No,Sun,Dinner,2,13.636364,4.840000,False
178,9.60,4.00,Female,Yes,Sun,Dinner,2,41.666667,4.800000,False
6,8.77,2.00,Male,No,Sun,Dinner,2,22.805017,4.385000,False


In [120]:
grupa = df.groupby("day")
print(grupa["tip_pct"].mean().reset_index())
#BEZ RESET INDEX CE DANI DA BUDU INDEKSI A NE POSEBNA KOLONA PA TO RADIMO DA BI OVO BILA KOLONA

    day    tip_pct
0  Thur  16.139743
1   Fri  16.991303
2   Sat  15.315172
3   Sun  16.689729


In [112]:
df.groupby("day").size()

day
Thur    61
Fri     19
Sat     87
Sun     76
dtype: int64

In [113]:
df.groupby("smoker")["tip_pct"].mean()

smoker
Yes    16.329767
No     15.932846
Name: tip_pct, dtype: float64

In [116]:
df.groupby(["day","time"]).size()

day   time  
Thur  Lunch     60
      Dinner     1
Fri   Lunch      7
      Dinner    12
Sat   Dinner    87
Sun   Dinner    76
dtype: int64

In [115]:
df.groupby("day", observed=True).agg(
    prosek_racun=("total_bill", "mean"),
    broj=("total_bill", "size"),
)

,prosek_racun,broj
day,,
Thur,17.759508,61
Fri,17.151579,19
Sat,20.441379,87
Sun,21.410000,76


In [121]:
print(df)

     total_bill   tip     sex smoker   day    time  size    tip_pct  \
0         16.99  1.01  Female     No   Sun  Dinner     2   5.944673   
1         10.34  1.66    Male     No   Sun  Dinner     3  16.054159   
2         21.01  3.50    Male     No   Sun  Dinner     3  16.658734   
3         23.68  3.31    Male     No   Sun  Dinner     2  13.978041   
4         24.59  3.61  Female     No   Sun  Dinner     4  14.680765   
..          ...   ...     ...    ...   ...     ...   ...        ...   
239       29.03  5.92    Male     No   Sat  Dinner     3  20.392697   
240       27.18  2.00  Female    Yes   Sat  Dinner     2   7.358352   
241       22.67  2.00    Male    Yes   Sat  Dinner     2   8.822232   
242       17.82  1.75    Male     No   Sat  Dinner     2   9.820426   
243       18.78  3.00  Female     No  Thur  Dinner     2  15.974441   

      po_osobi  velika_grupa  
0     8.495000         False  
1     3.446667         False  
2     7.003333         False  
3    11.840000         

In [124]:
df["total_bill"].value_counts()

total_bill
13.42    3
10.34    2
21.01    2
10.33    2
17.92    2
        ..
29.03    1
27.18    1
22.67    1
17.82    1
18.78    1
Name: count, Length: 229, dtype: int64

In [127]:
df["total_bill"].value_counts().sort_index(ascending=False)
#GORE SE SORTIRALO PO BROJU POJAVLJIVANJA A OVDE PO VREDNOSTI ZA TOTAL BILL

total_bill
50.81    1
48.33    1
48.27    1
48.17    1
45.35    1
        ..
7.56     1
7.51     1
7.25     2
5.75     1
3.07     1
Name: count, Length: 229, dtype: int64

In [128]:
df["total_bill"].value_counts(dropna = False)

total_bill
13.42    3
10.34    2
21.01    2
10.33    2
17.92    2
        ..
29.03    1
27.18    1
22.67    1
17.82    1
18.78    1
Name: count, Length: 229, dtype: int64

In [129]:
df[["sex", "smoker"]].value_counts()

sex     smoker
Male    No        97
        Yes       60
Female  No        54
        Yes       32
Name: count, dtype: int64

In [130]:
df.isnull().sum()

total_bill      0
tip             0
sex             0
smoker          0
day             0
time            0
size            0
tip_pct         0
po_osobi        0
velika_grupa    0
dtype: int64

## 2. Nedostajuće vrednosti
### Pronalaženje
python
t.isnull().sum()

isnull() pravi tabelu istog oblika, punu True/False, a sum() ih broji po kolonama. Isti trik sa jedinicama i nulama koji već znaš.

Za procenat koristiš mean umesto sum:

python
(t.isnull().mean() * 100).round(1)

U deck fali 77%, u age 20%, u embarked 0.2%. Procenat ti odmah govori šta da radiš: kolonu sa 77% praznih obično izbacuješ, dok se 0.2% popunjava bez razmišljanja.

### Popunjavanje
python
t["age"].fillna(t["age"].median())                # medijana
t["embarked"].fillna(t["embarked"].mode()[0])     # najčešća vrednost

Za age se koristi medijana, a ne prosek, jer je medijana otporna na ekstremne vrednosti. Ovde je medijana 28, a prosek 29.7.

mode() vraća Series, jer može biti više podjednako čestih vrednosti. Zato ide [0] da uzmeš prvu.

Pametnija verzija popunjava po grupama:

python
t["age"] = t["age"].fillna(
    t.groupby(["sex", "pclass"], observed=True)["age"].transform("median")
)

Ovde se spajaju stvari koje već znaš. groupby pravi grupe, a transform("median") vrati medijanu grupe nazad na svaki red te grupe. To je razlika u odnosu na mean(), koji bi vratio 6 brojeva. transform vraća 891, koliko i redova, pa se lepo poklapa sa fillna.

Rezultat ima smisla: muškarac iz prve klase dobija 40 godina, a devojka iz treće 21.5. Bolje nego da svi dobiju istih 28.

Izbacivanje
python
t.dropna()                      # svaki red u kome bilo šta fali
t.dropna(subset=["age"])        # samo gde fali starost
t.dropna(axis=1, thresh=800)    # kolone sa manje od 800 punih vrednosti

Pazi na prvi. Na Titanicu ti od 891 reda ostane samo 182, jer deck obori skoro sve. Zato se dropna() bez argumenata retko koristi. Sa subset=["age"] ostaje 714, što je već upotrebljivo.

### Zamka sa category kolonama
python
t["deck"].fillna("Unknown")

Baca grešku: Cannot setitem on a Categorical with a new category. Kolona je tipa category i zna samo za slova A do G. "Unknown" nije u spisku dozvoljenih.

Rešenje je da je prvo pretvoriš u običan tekst:

python
t["deck"] = t["deck"].astype(str).replace("nan", "Unknown")

Kad pretvoriš u str, NaN postane tekst "nan", pa ga menjaš kao običnu reč.

## 3. Menjanje vrednosti
map: tekst u broj preko rečnika
python
t["sex_num"] = t["sex"].map({"male": 0, "female": 1})

Ovo ti treba stalno, jer modeli ne razumeju reči. Pazi: vrednost koje nema u rečniku postaje NaN. Zato uvek posle proveri sa value_counts(dropna=False).

replace: zamena pojedinačnih vrednosti
python
t["embarked"].replace({"S": "Southampton", "C": "Cherbourg"})

Razlika u odnosu na map: replace ostavlja vrednosti koje nisi naveo, a map ih pretvara u NaN. Zato map koristiš kad prevodiš celu kolonu, a replace kad popravljaš par vrednosti.

astype: promena tipa
python
t["survived"].astype(bool)
t["fare"].astype(int)       # odseca decimale, ne zaokružuje
t["sex"].astype(str)
pd.cut: broj u kategorije
python
t["grupa"] = pd.cut(
    t["age"],
    bins=[0, 12, 18, 60, 100],
    labels=["dete", "tinejdzer", "odrastao", "stariji"],
)

Granice sam biraš. Imaj u vidu da uvek dobiješ jednu granicu više nego oznaka, jer četiri granice prave tri intervala, a pet granica četiri.

Odmah proveri da li nešto znači:

python
t.groupby("grupa", observed=True)["survived"].mean()

Deca 58%, tinejdžeri 43%, odrasli 39%, stariji 23%. Jasan obrazac.

Postoji i pd.qcut, koji sam bira granice tako da u svakoj grupi bude podjednako redova. Koristan za cene karata, gde ti granice nisu očigledne.

.loc za dodelu

Kad menjaš vrednosti samo u nekim redovima, ide .loc sa maskom i imenom kolone:

python
t.loc[t["fare"] > 200, "skupa_karta"] = 1

Čita se: u redovima gde je cena preko 200, u koloni skupa_karta upiši 1.

Nikad ne piši ovako:

python
t[t["fare"] > 200]["skupa_karta"] = 1   # ne radi

Ovo prvo napravi kopiju, upiše u kopiju, pa je baci. Original ostane nepromenjen, a ti dobiješ SettingWithCopyWarning ili tiho ništa. To je ona ista priča o copy i view iz zadatka N5.

np.where: uslov u dve vrednosti
python
t["punoletan"] = np.where(t["age"] < 18, "maloletan", "punoletan")

Čita se: gde je uslov tačan stavi prvo, inače drugo. Kad ti treba više od dva ishoda, koristi se np.select ili pd.cut.

In [134]:
import numpy as np
rng = np.random.default_rng(42)
df.loc[rng.choice(df.index, 30, replace=False), "total_bill"] = np.nan
df.loc[rng.choice(df.index, 12, replace=False), "size"] = np.nan
df.loc[rng.choice(df.index, 60, replace=False), "day"] = np.nan

In [137]:
df.value_counts("day")

day
Sat     64
Sun     59
Thur    44
Fri     16
Name: count, dtype: int64

In [138]:
df.value_counts("day", dropna=False)

day
Sat     64
NaN     60
Sun     59
Thur    44
Fri     16
Name: count, dtype: int64

In [139]:
df.value_counts("smoker", normalize = True)

smoker
No     0.621399
Yes    0.378601
Name: proportion, dtype: float64

In [140]:
df.value_counts("smoker")

smoker
No     151
Yes     92
Name: count, dtype: int64

In [145]:
df.value_counts("size", normalize = True).sort_index(ascending = False)

size
6.0    0.017316
5.0    0.021645
4.0    0.155844
3.0    0.160173
2.0    0.627706
1.0    0.017316
Name: proportion, dtype: float64

### Za ovo iznad sve znaci imamo par stvari
- value_counts() se koristi za brojanje postojecih vrednosti znaci on nan ne racuna, ako ocemo da prikazemo i koliko ih je nan kazemo u zagradi **dropna = False**. U zagradi stavljamo prvo za koju kolonu prebrojavamo vrednosti, ako hocemo da imam procente onda kazemo u zagradi **normalize = True**
- sort_index() se koristi da se sortira po vrednostima a ne po broju ponavljanja tj frekv tih vrednosti, ako hocemo opadajuce onda stavimo **ascending=False**
- reset_index() se koristi posto nam value_counts vrati Series da to pretvorimo u dataframe odnosno indeksi su nam ustvari vrednosti neke kolone u Series a onda to postaje kolona pored druge kolone koja je frekvenca i to je Dataframe sada sa 2 kolone a indeksi su 0 1 2 itd

In [161]:
(df.isna().sum()).sort_values(ascending = False)

day             60
total_bill      30
size            12
tip              0
smoker           0
sex              0
time             0
tip_pct          0
po_osobi         0
velika_grupa     0
dtype: int64

In [157]:
((df.isnull().sum()/len(df)) * 100).round(1)

total_bill      12.3
tip              0.0
sex              0.0
smoker           0.0
day             24.7
time             0.0
size             4.9
tip_pct          0.0
po_osobi         0.0
velika_grupa     0.0
dtype: float64

In [165]:
df["total_bill"].median().round(2)

np.float64(17.47)

In [166]:
df["total_bill"].mean().round(2)

np.float64(19.74)

### Zadatak: Popuni total_bill medijanom po vremenu obroka, preko groupby i transform.
Znaci krecemo sa dodelom to je logicno posto popunjavamo kolonu, fillna sluzi da popunimo samo vrednosti koje fale, e sad dalje groupby nam je sluzio da grupisemo sve po vremenu za total bill a transform je da se vrate svi redovi odnosno da ne bude jedna vrednost tipa za veceru nego za svaku kolonu gde je vecera ponavljamo tu vrednost jer tako imamo broj redova koliko i u standardnom df pa ce zato da se koristi transform da bi se samo gdee nema vrednost upisala vrednost za taj red iz rezultata

In [170]:
df["total_bill"] = df["total_bill"].fillna(df.groupby("time")["total_bill"].transform("median"))

In [171]:
print(df)

     total_bill   tip     sex smoker   day    time  size    tip_pct  \
0         16.99  1.01  Female     No   Sun  Dinner   2.0   5.944673   
1         10.34  1.66    Male     No   NaN  Dinner   3.0  16.054159   
2         21.01  3.50    Male     No   Sun  Dinner   3.0  16.658734   
3         23.68  3.31    Male     No   Sun  Dinner   2.0  13.978041   
4         24.59  3.61  Female     No   Sun  Dinner   4.0  14.680765   
..          ...   ...     ...    ...   ...     ...   ...        ...   
239       29.03  5.92    Male     No   Sat  Dinner   3.0  20.392697   
240       27.18  2.00  Female    Yes   NaN  Dinner   2.0   7.358352   
241       22.67  2.00    Male    Yes   NaN  Dinner   2.0   8.822232   
242       17.82  1.75    Male     No   Sat  Dinner   2.0   9.820426   
243       18.78  3.00  Female     No  Thur  Dinner   2.0  15.974441   

      po_osobi  velika_grupa  
0     8.495000         False  
1     3.446667         False  
2     7.003333         False  
3    11.840000         

In [172]:
df.value_counts("total_bill")

total_bill
18.04    22
15.98    11
13.42     3
10.34     2
21.01     2
         ..
29.03     1
27.18     1
22.67     1
17.82     1
18.78     1
Name: count, Length: 202, dtype: int64

In [173]:
df["total_bill"].mean()

np.float64(19.454732510288064)

In [175]:
df.dropna()

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
0,16.99,1.01,Female,No,Sun,Dinner,2.0,5.944673,8.495000,False
2,21.01,3.50,Male,No,Sun,Dinner,3.0,16.658734,7.003333,False
3,23.68,3.31,Male,No,Sun,Dinner,2.0,13.978041,11.840000,False
4,24.59,3.61,Female,No,Sun,Dinner,4.0,14.680765,6.147500,True
5,25.29,4.71,Male,No,Sun,Dinner,4.0,18.623962,6.322500,True
...,...,...,...,...,...,...,...,...,...,...
236,12.60,1.00,Male,Yes,Sat,Dinner,2.0,7.936508,6.300000,False
237,18.04,1.17,Male,Yes,Sat,Dinner,2.0,3.563814,16.415000,False
239,29.03,5.92,Male,No,Sat,Dinner,3.0,20.392697,9.676667,False
242,17.82,1.75,Male,No,Sat,Dinner,2.0,9.820426,8.910000,False


In [176]:
df.dropna(subset = ["total_bill"])

,total_bill,tip,sex,smoker,day,time,size,tip_pct,po_osobi,velika_grupa
0,16.99,1.01,Female,No,Sun,Dinner,2.0,5.944673,8.495000,False
1,10.34,1.66,Male,No,NaN,Dinner,3.0,16.054159,3.446667,False
2,21.01,3.50,Male,No,Sun,Dinner,3.0,16.658734,7.003333,False
3,23.68,3.31,Male,No,Sun,Dinner,2.0,13.978041,11.840000,False
4,24.59,3.61,Female,No,Sun,Dinner,4.0,14.680765,6.147500,True
...,...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3.0,20.392697,9.676667,False
240,27.18,2.00,Female,Yes,NaN,Dinner,2.0,7.358352,13.590000,False
241,22.67,2.00,Male,Yes,NaN,Dinner,2.0,8.822232,11.335000,False
242,17.82,1.75,Male,No,Sat,Dinner,2.0,9.820426,8.910000,False


Ovde ima nesto nzm ni ja koja je fora uglavnmom kategoricke su vrednosti dani u nedelji i ne dozvoljava da dodas novu vrednost pa ce onda ovo astype(str) da promeni to u string i onda moze da se koristi tj da se dodaju vrednosti kao sto je Nepoznat

In [179]:
df["day"].astype(str).fillna("Nepoznat")

0           Sun
1      Nepoznat
2           Sun
3           Sun
4           Sun
         ...   
239         Sat
240    Nepoznat
241    Nepoznat
242         Sat
243        Thur
Name: day, Length: 243, dtype: str

In [182]:
df["size"] = df["size"].fillna(df["size"].median()).astype(int)
#KADA MENJAMO TIP OVAKO SE SAMO DODA ASTYPE

In [197]:
d = sns.load_dataset("tips").copy()
#OVO COPY JE DA SE SLUCAJNO NE DESI POSTO SMO VEC UCITAVALI DATASET ISTI DA NAM DA SAMO REFERENCU NA TO STO IMAMO UCITANO TJ NA IZMENJENI

In [198]:
d["smoker"] = d["smoker"].map({"Yes":1, "No" : 0})

In [199]:
d

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,0,Sun,Dinner,2
1,10.34,1.66,Male,0,Sun,Dinner,3
2,21.01,3.50,Male,0,Sun,Dinner,3
3,23.68,3.31,Male,0,Sun,Dinner,2
4,24.59,3.61,Female,0,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,0,Sat,Dinner,3
240,27.18,2.00,Female,1,Sat,Dinner,2
241,22.67,2.00,Male,1,Sat,Dinner,2
242,17.82,1.75,Male,0,Sat,Dinner,2


In [200]:
d["sex"] = d["sex"].map({"Female" : 1, "Male" : 0})
d

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,1,0,Sun,Dinner,2
1,10.34,1.66,0,0,Sun,Dinner,3
2,21.01,3.50,0,0,Sun,Dinner,3
3,23.68,3.31,0,0,Sun,Dinner,2
4,24.59,3.61,1,0,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3
240,27.18,2.00,1,1,Sat,Dinner,2
241,22.67,2.00,0,1,Sat,Dinner,2
242,17.82,1.75,0,0,Sat,Dinner,2


In [201]:
d["smoker"].value_counts(dropna = False)

smoker
0    151
1     93
Name: count, dtype: int64

In [202]:
d["sex"].value_counts(dropna = False)

sex
0    157
1     87
Name: count, dtype: int64

In [204]:
d["day"] = d["day"].astype(str).replace({"Thur" : "Cetvrtak", "Fri" : "Petak"})

In [205]:
d

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,1,0,Sun,Dinner,2
1,10.34,1.66,0,0,Sun,Dinner,3
2,21.01,3.50,0,0,Sun,Dinner,3
3,23.68,3.31,0,0,Sun,Dinner,2
4,24.59,3.61,1,0,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3
240,27.18,2.00,1,1,Sat,Dinner,2
241,22.67,2.00,0,1,Sat,Dinner,2
242,17.82,1.75,0,0,Sat,Dinner,2


In [208]:
d["tip_pct"] = (d.tip / d.total_bill * 100).round(2)

In [209]:
d

,total_bill,tip,sex,smoker,day,time,size,tip_pct
0,16.99,1.01,1,0,Sun,Dinner,2,5.94
1,10.34,1.66,0,0,Sun,Dinner,3,16.05
2,21.01,3.50,0,0,Sun,Dinner,3,16.66
3,23.68,3.31,0,0,Sun,Dinner,2,13.98
4,24.59,3.61,1,0,Sun,Dinner,4,14.68
...,...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3,20.39
240,27.18,2.00,1,1,Sat,Dinner,2,7.36
241,22.67,2.00,0,1,Sat,Dinner,2,8.82
242,17.82,1.75,0,0,Sat,Dinner,2,9.82


In [213]:
d["bill_grupa"] = pd.cut(d["total_bill"], bins = [0, 15, 25, 40, 60], labels = ["mali", "srednji", "veliki", "ogroman"])

In [214]:
d


,total_bill,tip,sex,smoker,day,time,size,tip_pct,bill_grupa
0,16.99,1.01,1,0,Sun,Dinner,2,5.94,srednji
1,10.34,1.66,0,0,Sun,Dinner,3,16.05,mali
2,21.01,3.50,0,0,Sun,Dinner,3,16.66,srednji
3,23.68,3.31,0,0,Sun,Dinner,2,13.98,srednji
4,24.59,3.61,1,0,Sun,Dinner,4,14.68,srednji
...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3,20.39,veliki
240,27.18,2.00,1,1,Sat,Dinner,2,7.36,veliki
241,22.67,2.00,0,1,Sat,Dinner,2,8.82,srednji
242,17.82,1.75,0,0,Sat,Dinner,2,9.82,srednji


In [215]:
d["bill_grupa"].value_counts()

bill_grupa
srednji    109
mali        80
veliki      45
ogroman     10
Name: count, dtype: int64

In [218]:
d.groupby("bill_grupa")["tip_pct"].mean() #agg("mean")

bill_grupa
mali       18.349125
srednji    15.882385
veliki     13.454667
ogroman    11.889000
Name: tip_pct, dtype: float64

In [225]:
pd.qcut(d["total_bill"], 4, labels = ["mali", "srednji", "veliki", "ogroman"]).value_counts().sort_index()

total_bill
mali       61
srednji    61
veliki     61
ogroman    61
Name: count, dtype: int64

In [227]:
d["tip_kat"] = "Normalna"

In [229]:
d["tip_kat"] = pd.cut(d["tip_pct"], bins=[0, 10, 20, 100], labels=["Skrta", "Normalna", "Izdasna"])

In [230]:
d

,total_bill,tip,sex,smoker,day,time,size,tip_pct,bill_grupa,tip_kat
0,16.99,1.01,1,0,Sun,Dinner,2,5.94,srednji,Skrta
1,10.34,1.66,0,0,Sun,Dinner,3,16.05,mali,Normalna
2,21.01,3.50,0,0,Sun,Dinner,3,16.66,srednji,Normalna
3,23.68,3.31,0,0,Sun,Dinner,2,13.98,srednji,Normalna
4,24.59,3.61,1,0,Sun,Dinner,4,14.68,srednji,Normalna
...,...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3,20.39,veliki,Izdasna
240,27.18,2.00,1,1,Sat,Dinner,2,7.36,veliki,Skrta
241,22.67,2.00,0,1,Sat,Dinner,2,8.82,srednji,Skrta
242,17.82,1.75,0,0,Sat,Dinner,2,9.82,srednji,Skrta


In [231]:
d[d["tip_pct"] > 20]["tip_kat"] = "Izdasna"
d

C:\Users\Aleksandar\AppData\Local\Temp\ipykernel_9756\937957902.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment.
Such chained assignment never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

Try using '.loc[row_indexer, col_indexer] = value' instead, to perform the assignment in a single step.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html#chained-assignment
  d[d["tip_pct"] > 20]["tip_kat"] = "Izdasna"


,total_bill,tip,sex,smoker,day,time,size,tip_pct,bill_grupa,tip_kat
0,16.99,1.01,1,0,Sun,Dinner,2,5.94,srednji,Skrta
1,10.34,1.66,0,0,Sun,Dinner,3,16.05,mali,Normalna
2,21.01,3.50,0,0,Sun,Dinner,3,16.66,srednji,Normalna
3,23.68,3.31,0,0,Sun,Dinner,2,13.98,srednji,Normalna
4,24.59,3.61,1,0,Sun,Dinner,4,14.68,srednji,Normalna
...,...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3,20.39,veliki,Izdasna
240,27.18,2.00,1,1,Sat,Dinner,2,7.36,veliki,Skrta
241,22.67,2.00,0,1,Sat,Dinner,2,8.82,srednji,Skrta
242,17.82,1.75,0,0,Sat,Dinner,2,9.82,srednji,Skrta


In [233]:
d["sto_kat"] = np.where(d["size"] >= 4, "Velika", "Mala")
d

,total_bill,tip,sex,smoker,day,time,size,tip_pct,bill_grupa,tip_kat,sto_kat
0,16.99,1.01,1,0,Sun,Dinner,2,5.94,srednji,Skrta,Mala
1,10.34,1.66,0,0,Sun,Dinner,3,16.05,mali,Normalna,Mala
2,21.01,3.50,0,0,Sun,Dinner,3,16.66,srednji,Normalna,Mala
3,23.68,3.31,0,0,Sun,Dinner,2,13.98,srednji,Normalna,Mala
4,24.59,3.61,1,0,Sun,Dinner,4,14.68,srednji,Normalna,Velika
...,...,...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3,20.39,veliki,Izdasna,Mala
240,27.18,2.00,1,1,Sat,Dinner,2,7.36,veliki,Skrta,Mala
241,22.67,2.00,0,1,Sat,Dinner,2,8.82,srednji,Skrta,Mala
242,17.82,1.75,0,0,Sat,Dinner,2,9.82,srednji,Skrta,Mala


In [234]:
d.groupby("sto_kat")["tip_pct"].mean()

sto_kat
Mala      16.415202
Velika    14.635870
Name: tip_pct, dtype: float64

In [235]:
d["vikend"] = d["day"].isin(["Sat", "Sun"])

In [236]:
d

,total_bill,tip,sex,smoker,day,time,size,tip_pct,bill_grupa,tip_kat,sto_kat,vikend
0,16.99,1.01,1,0,Sun,Dinner,2,5.94,srednji,Skrta,Mala,True
1,10.34,1.66,0,0,Sun,Dinner,3,16.05,mali,Normalna,Mala,True
2,21.01,3.50,0,0,Sun,Dinner,3,16.66,srednji,Normalna,Mala,True
3,23.68,3.31,0,0,Sun,Dinner,2,13.98,srednji,Normalna,Mala,True
4,24.59,3.61,1,0,Sun,Dinner,4,14.68,srednji,Normalna,Velika,True
...,...,...,...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,0,0,Sat,Dinner,3,20.39,veliki,Izdasna,Mala,True
240,27.18,2.00,1,1,Sat,Dinner,2,7.36,veliki,Skrta,Mala,True
241,22.67,2.00,0,1,Sat,Dinner,2,8.82,srednji,Skrta,Mala,True
242,17.82,1.75,0,0,Sat,Dinner,2,9.82,srednji,Skrta,Mala,True


In [238]:
d.groupby("vikend")["tip_pct"].mean().round(2)

vikend
False    16.33
True     15.96
Name: tip_pct, dtype: float64